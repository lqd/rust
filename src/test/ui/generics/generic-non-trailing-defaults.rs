struct Heap;

struct Vec<A = Heap, T>(A, T);
//~^ ERROR type parameters must be declared prior to defaulted parameters

struct Foo<A, B = Vec<C>, C>(A, B, C);
//~^ ERROR type parameters must be declared prior to defaulted parameters
//~| ERROR type parameters with a default cannot use forward declared identifiers

fn main() {}
