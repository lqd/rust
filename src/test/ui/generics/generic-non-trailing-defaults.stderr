error: type parameters must be declared prior to defaulted parameters
  --> $DIR/generic-non-trailing-defaults.rs:3:22
   |
LL | struct Vec<A = Heap, T>(A, T);
   |           -----------^- help: reorder the parameters: `<T, A = Heap>`
   |
   = note: the expected order for parameters is: lifetimes, then types, then consts, then defaulted parameters

error: type parameters must be declared prior to defaulted parameters
  --> $DIR/generic-non-trailing-defaults.rs:6:27
   |
LL | struct Foo<A, B = Vec<C>, C>(A, B, C);
   |           ----------------^- help: reorder the parameters: `<A, C, B = Vec<C>>`
   |
   = note: the expected order for parameters is: lifetimes, then types, then consts, then defaulted parameters

error[E0128]: type parameters with a default cannot use forward declared identifiers
  --> $DIR/generic-non-trailing-defaults.rs:6:23
   |
LL | struct Foo<A, B = Vec<C>, C>(A, B, C);
   |                       ^ defaulted type parameters cannot be forward declared

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0128`.
